bigtable-dataflow-parent/bigtable-beam-import/src/test/java/com/google/cloud/bigtable/beam/hbasesnapshots/EndToEndIT.java

```diff
@@ -185,6 +245,81 @@ public void testHBaseSnapshotImport() throws Exception {
+    state = result.waitUntilFinish();
+    Assert.assertEquals(State.DONE, state);
+
+    List<GcsPath> outputs = gcsUtil.expand(GcsPath.fromUri(syncTableOutputDir + "*"));
+    // FileSink will write an empty file when there are no mismatches
+    Assert.assertEquals(1, outputs.size());
+    // TODO read the actual files and validate the ranges instead of size check
+    Assert.assertEquals(0, gcsUtil.fileSize(outputs.get(0)));
+
+    // Validate the counters.
+    Map<String, Long> counters = getCountMap(result);
+    Assert.assertEquals(counters.size(), 1);
+    Assert.assertEquals(counters.get("ranges_matched"), (Long) 101L);
+  }
+
+  /**
+   * Introduces multiple corruptions in imported table and validates that sync-table can detect
+   * them.
+   */
+  @Test
+  public void testHBaseSnapshotImportWithCorruptions() throws Exception {
+    // Import snapshot
+    ImportOptions importOpts = createImportOptions();
+    State state = ImportJobFromHbaseSnapshot.buildPipeline(importOpts).run().waitUntilFinish();
+    Assert.assertEquals(State.DONE, state);
+
+    // Introduce corruptions to the data in Bigtable. Delete data from Bigtable to simulate Bigtable
+    // missing data. Add data to Bigtable to simulate extra data in Bigtable. It is easier to update
+    // Bigtable than change the snapshots.
+    Table table = connection.getTable(TableName.valueOf(tableId));
+    Cell cellInMiddle = table.get(new Get("24".getBytes())).rawCells()[0];
+    List<Put> puts =
+        Arrays.asList(
+            // Add a row at the start
+            new Put(Bytes.toBytes("000"))
+                .addColumn(CF.getBytes(), "random_col".getBytes(), 1L, "value000".getBytes())
+                .addColumn(CF.getBytes(), "random_col".getBytes(), 2L, "value001".getBytes()),
+            // change a cell in middle
+            new Put(cellInMiddle.getRowArray())
+                .addColumn(
+                    cellInMiddle.getFamilyArray(),
+                    cellInMiddle.getQualifierArray(),
+                    cellInMiddle.getTimestamp(),
+                    "corrupted_val".getBytes()),
+            // add a new row in the end
+            new Put(Bytes.toBytes("9999"))
+                .addColumn(CF.getBytes(), "random_col".getBytes(), 100L, "value999".getBytes()));
+
+    table.put(puts);
+    // Delete a random row in the middle. We should see 4 ranges mismatch as table is split on
+    // 1,2...9. We are splitting on 31, delete in 60s.
+    table.delete(new Delete("64".getBytes()));
+
+    // Run SyncTable job and expect 4 mismatches.
+    SyncTableOptions syncOpts = createSyncTableOptions();
+    PipelineResult result = SyncTableJob.buildPipeline(syncOpts).run();
+    state = result.waitUntilFinish();
+    Assert.assertEquals(State.DONE, state);
+
+    List<GcsPath> outputs = gcsUtil.expand(GcsPath.fromUri(syncTableOutputDir + "*"));
+
+    System.out.println("OUTPUTS: " + outputs);
```

Comment:
Please avoid using sys.out. Please use logging instead or better yet assert that the outputs match expected output

Reply:
Yes, we will assert them. Need to have a conversation with you on the output format.